### PR TITLE
Handle newer docker remote API pulling from older/legacy registry

### LINKED ISSUE #380: Legacy registry message on pull results in error status
When using a newer docker version to pull from a v1 docker registry, the docker API can respond with

> registry/image: this image was pulled from a legacy registry.  Important: This registry version will not be supported in future versions of docker.

This occurs even though the pull was successful to the remote API. When this happens, `PullResponseItem` does not recognize this status message as successful.

### PR DESCRIPTION
**Pull request by zachmarshall**
Adds support for handling status message referencing legacy registry. Also adds tests for changes and existing behavior.

Fixes #380

### CODE REVIEW

File: src/main/java/com/github/dockerjava/api/model/PullResponseItem.java [hunk 1]
```diff
@@ -22,8 +22,9 @@ public boolean isPullSuccessIndicated() {
             return false;
         }
 
-        return (getStatus().contains("Download complete") || getStatus().contains("Image is up to date") || getStatus()
-                .contains("Downloaded newer image"));
+        return (getStatus().contains("Download complete") || getStatus().contains("Image is up to date")
+                || getStatus().contains("Downloaded newer image")
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/PullResponseItem.java]: Please extract messages to public static final

File: src/test/java/com/github/dockerjava/api/model/PullResponseItemTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+import org.testng.annotations.Test;
+
+/**
+ * Tests logic of PullResponseItem's error/success handling by simulating a JSON response.
+ *
+ * @author Zach Marshall
+ */
+public class PullResponseItemTest
+{
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/api/model/PullResponseItemTest.java]: { on the same line as class declaration

File: src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java [hunk 3]
```diff
@@ -45,7 +44,7 @@
      *             JSON Conversion error
      */
     public static String readString(JSONResourceRef resource) throws IOException {
-        InputStream istream = CommandJSONSamples.class.getResourceAsStream(resource.getFileName());
+        InputStream istream = resource.getResourceClass().getResourceAsStream(resource.getFileName());
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java]: wrap with try-with-resources

File: src/test/resources/com/github/dockerjava/api/model/pullImageResponse_error.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"errorDetail":{"message":"Error: image cccxxx/xxxccc:latest not found"},"error":"Error: image cccxxx/xxxccc:latest not found"}
```
Comment (KostyaSha) [inline src/test/resources/com/github/dockerjava/api/model/pullImageResponse_error.json]: add new lines

File: src/test/java/com/github/dockerjava/api/model/PullResponseJSONSamples.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import com.github.dockerjava.test.serdes.JSONResourceRef;
+
+/**
+ * Enumeration for the available pull response statuses.
+ *
+ * @author Zach Marshall
+ */
+public enum PullResponseJSONSamples implements JSONResourceRef
+{
+    pullImageResponse_legacy, pullImageResponse_error, pullImageResponse_newerImage, pullImageResponse_upToDate;
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/api/model/PullResponseJSONSamples.java]: @marcuslinke TODO create helper that just loads jsons from simple folders structure "String type, String ver, String jsonName" where type will be registry|| engine, ver - some ver i.e. 1.23 and jsonName real file.

**Comment by KostyaSha** [comment]
Please mark in commit message from what docker registry to what docker host it was connected

**Comment by zachmarshall** [comment]
Hi @KostyaSha thanks for the feedback, I will make those changes.

Regarding the commit message registry/host. Are you asking for which version of the registry and which version of the docker host? Or are you asking for the hostname of the registry and api that I found this issue? The issue was found with a private docker registry and API so I wouldn't be able to provide the hostnames.

File: src/test/java/com/github/dockerjava/api/model/PullResponseJSONSamples.java (hunk 5 above)
Comment (zachmarshall) [inline src/test/java/com/github/dockerjava/api/model/PullResponseJSONSamples.java]: Yeah I don't like this enum stuff either, just following what was already there....

**Comment by zachmarshall** [comment]
Hi @KostyaSha, any update on the feedback regarding the commit message? If not, are you ready to merge this?

Also, do you want another PR for master/version 3 or will you merge the 2.x branch and this fix yourselves?

**Comment by KostyaSha** [comment]
@zachmarshall sorry, was AFK (travelling) back to home today. I have some number of enhacement locally will try get some window to resolve portion of PRs and push local changes! Sorry for delays.

File: src/main/java/com/github/dockerjava/api/model/PullResponseItem.java [hunk 6]
```diff
@@ -22,8 +26,9 @@ public boolean isPullSuccessIndicated() {
             return false;
         }
 
-        return (getStatus().contains("Download complete") || getStatus().contains("Image is up to date") || getStatus()
-                .contains("Downloaded newer image"));
+        return (getStatus().contains(DOWNLOAD_COMPLETE) || getStatus().contains(IMAGE_UP_TO_DATE)
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/api/model/PullResponseItem.java]: Hm.. will it make sense organise to some better structure instead ifs?

File: src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java [hunk 7]
```diff
@@ -25,14 +25,14 @@
...
 import com.fasterxml.jackson.databind.ObjectMapper;
-import com.github.dockerjava.api.command.CommandJSONSamples;
 
 /**
  * Provides helper methods for serialization-deserialization tests
  * 
  * @author Oleg Nenashev
- * @since TODO
  */
+//TODO: Create helper that loads json files from simple folder structure using a type,
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java]: better place in javadoc, more chances that somebody will see it :) Though i plan change it soon.

File: src/test/resources/com/github/dockerjava/api/model/pullImageResponse_error.json (hunk 4 above)
Comment (KostyaSha) [inline src/test/resources/com/github/dockerjava/api/model/pullImageResponse_error.json]: Thanks

File: src/main/java/com/github/dockerjava/api/model/PullResponseItem.java (hunk 6 above)
Comment (zachmarshall) [inline src/main/java/com/github/dockerjava/api/model/PullResponseItem.java]: Not sure I understand. What were you envisioning? Using explicit if statements? Not sure if that is clearer.

Besides, I am trying to minimize the size of this commit to just fix the issue identified. I am happy to help with code cleanup and formatting in another issue/PR if that is useful but that seems outside the scope of this change.

File: src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java (hunk 7 above)
Comment (zachmarshall) [inline src/test/java/com/github/dockerjava/test/serdes/JSONTestHelper.java]: Will do.

**Comment by zachmarshall** [comment]
I see you added this to 3.0.0 milestone. Note that this PR is for the 2.x branch since that is what the downstream project I am using (gradle-docker-plugin) uses. I can submit a 3.0.0 PR if you'd like, or you can merge this branch.

**Comment by KostyaSha** [comment]
Could you open parallel pr to 3.0.0 to not lost changes?

File: src/test/java/com/github/dockerjava/api/model/PullResponseItemTest.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        PullResponseItem response =
+                testRoundTrip(PullResponseJSONSamples.pullImageResponse_upToDate, PullResponseItem.class);
+        assertTrue(response.isPullSuccessIndicated());
+        assertFalse(response.isErrorIndicated());
+    }
+
+    @Test
+    public void pullLegacyRegistry()
+        throws IOException
+    {
```
Comment (KostyaSha) [inline src/test/java/com/github/dockerjava/api/model/PullResponseItemTest.java]: `{` without new lines

**Comment by zachmarshall** [comment]
Opened PR #385 for parallel changes to 3.0.0/master.

**Comment by marcuslinke** [comment]
@KostyaSha @zachmarshall Wouldn't it be better to merge commits from 2.x branch to master?

**Comment by KostyaSha** [comment]
@marcuslinke i expect conflicts... as you wish, but then i will be able mark milestones.

**Comment by zachmarshall** [comment]
@marcuslinke @KostyaSha I am fine to do either, up to you guys obviously :)

**Comment by marcuslinke** [comment]
@KostyaSha I'm not really sure how to deal best with it. I guess it would be better to merge 2.x commit into master manually but if this feels to expensive to you I'm OK with the doubled PRs also.

**Comment by KostyaSha** [comment]
Picked via https://github.com/docker-java/docker-java/commit/c03f0c5ed2dc54248113f36a9a70843dd671ce30 (github didn't recognised change automatically).
